analysis/README.md

```diff
@@ -0,0 +1,623 @@
+| latch                   | no      | yes         |
+| function_id             | no      | yes         |
+| functions bodies        | yes     | no          |
+| internal instructions   | yes     | no          |
+
+The pipeline accepts both kinds of machines, and they are represented by the same type `Machine`. Some steps are specific to virtual machines. They can still be applied to constrained machines and must have no effect. In the process, virtual machines get reduced to constrained machines by encoding their high-level elements into constrained machines elements.
+
+## Pipeline
+
+We go through the pipeline using the following program as an example:
+
+```
+machine Main {
+
+    degree 16;
+
+    DifferentSignatures sub;
+
+    reg pc[@pc];
+    reg X[<=];
+    reg Y[<=];
+
+    instr identity X -> Y = sub.identity
+    instr one -> Y = sub.one
+    instr nothing = sub.nothing
+
+    function main {
+        nothing;
+        return;
+    }
+}
+
+// A machine exposing functions of different signatures
+machine DifferentSignatures {
+
+    reg pc[@pc];
+
+    function identity x: field -> field {
+        return x;
+    }
+
+    function one -> field {
+        return 1;
+    }
+
+    function nothing {
+        return;
+    }
+}
+```
+
+### Type checking
+
+Type checking takes a parse tree and returns a tree of machines. The output type aims at being as strict as possible.
+
+### Virtual machine reduction
+
+Virtual machine reduction turns virtual machines into constrained machines. It has no effect on contrained machines.
+
+#### Inference
+
+Inference infers the value of assignment registers.
+
+```diff
+-               A <== one();
++               A <=Y= one();
+```
+
+#### Batcher
+
+The batcher groups statements within each function into batches which can be executed in the same execution step. In our example, we gain 33% in the main machine by batching the first label with the following statement.
+
+```diff
++               // END BATCH
++               // END BATCH
++               // END BATCH
++               // END BATCH Unimplemented
++               // END BATCH
+```
+
+#### ASM to PIL
+
+ASM to PIL has two steps: ROM generation and reduction to constrained.
+
+##### ROM generation
+
+Rom generation generates a single ROM for each virtual machine using the following process:
+- Find the maximum number of inputs among all functions. Introduce as many input registers. Do the same for outputs, introducing output registers
+- Replace references to the function arguments by references to these input registers.
+- Pad all return statements with zeroes up to the number of output registers.
+- Inline all functions, adding a label before each one.
+- Add an infinite loop also behind a label
+- Enable non-deterministically jumping to one of these labels by setting `function_id` to the index of the label
+
+This process introduces instructions and registers, and returns the modified machine along with the ROM.
+
```

Comment:
Can you also add an example of rom gen with the dispatcher?

Reply:
Added the two roms for the example

Reply:
nice